File: src/backend/src/product/product.repository.ts
import { Injectable } from '@nestjs/common';
import { Between, Brackets, DataSource, EntityManager, In, Repository, Not } from 'typeorm';
import { ProductLog } from './entities/product-audit-log.entity';
import { Product } from './entities/product.entity';

@Injectable()
export class ProductRepository {
  private repository: Repository<Product>;
  private auditLogRepository: Repository<ProductLog>;

  constructor(private dataSource: DataSource) {
    this.repository = this.dataSource.getRepository(Product);
    this.auditLogRepository = this.dataSource.getRepository(ProductLog);
  }

  async findById(id: number): Promise<Product | null> {
    return this.repository.findOne({
      where: { productID: id, status: Not('DELETED') }
    });
  }

  async getAvailableStock(id: number): Promise<number> {
    const product = await this.findById(id);
    return product ? product.quantityInStock : 0;
  }

  async findProductsByIds(ids: number[]): Promise<Product[]> {
    if (!ids.length) {
      return [];
    }

    return this.repository.find({
      where: { productID: In(ids), status: Not('DELETED') }
    });
  }

  async searchProducts(
    keyword?: string,
    category?: string,
    minPrice?: number,
    maxPrice?: number,
    mediaTypes?: string[],
    status?: string,
  ): Promise<Product[]> {
    const query = this.repository
      .createQueryBuilder('product')
      .leftJoin('books', 'book', 'book.product_id = product.product_id')
      .leftJoin('cds', 'cd', 'cd.product_id = product.product_id')
      .leftJoin('cd_tracks', 'cd_track', 'cd_track.product_id = product.product_id')
      .leftJoin('dvds', 'dvd', 'dvd.product_id = product.product_id')
      .leftJoin('newspapers', 'newspaper', 'newspaper.product_id = product.product_id')
      .leftJoin('media', 'media', 'media.product_id = product.product_id');
    
    if (status) {
      if (status.toUpperCase() !== 'ALL') {
        query.where('product.status = :status', { status: status.toUpperCase() });
      } else {
        query.where('product.status != :status', { status: 'DELETED' });
      }
    } else {
      query.where('product.status = :status', { status: 'ACTIVE' });
    }

    const normalizedKeyword = keyword?.trim();
    if (normalizedKeyword) {
      query.andWhere(
        new Brackets((qb) => {
          qb.where('product.title ILIKE :keyword')
            .orWhere('product.category ILIKE :keyword')
            .orWhere('product.description ILIKE :keyword')
            .orWhere('product.barcode ILIKE :keyword')
            .orWhere('book.authors ILIKE :keyword')
            .orWhere('cd.artists ILIKE :keyword')
            .orWhere('cd_track.title ILIKE :keyword')
            .orWhere('dvd.director ILIKE :keyword')
            .orWhere('newspaper.editor_in_chief ILIKE :keyword')
            .orWhere('newspaper.sections ILIKE :keyword')
            .orWhere('media.publisher ILIKE :keyword')
            .orWhere('media.genre ILIKE :keyword')
            .orWhere('media.language ILIKE :keyword');
        }),
        { keyword: `%${normalizedKeyword}%` },
      );
    }
    if (category) {
      const categoryAliases = this.getCategoryAliases(category);
      const mediaType = this.getMediaTypeForCategory(category);

      query.andWhere(
        new Brackets((qb) => {
          qb.where('product.category IN (:...categoryAliases)', {
            categoryAliases,
          });

          if (mediaType) {
            qb.orWhere('product.product_type = :mediaType', { mediaType });
          }
        }),
      );
    }
    if (mediaTypes?.length) {
      query.andWhere('product.product_type IN (:...mediaTypes)', {
        mediaTypes,
      });
    }
    if (minPrice !== undefined) {
      query.andWhere('product.current_price >= :minPrice', { minPrice });
    }
    if (maxPrice !== undefined) {
      query.andWhere('product.current_price <= :maxPrice', { maxPrice });
    }

    return query.distinct(true).orderBy('product.title', 'ASC').getMany();
  }

  async findRandomActive(limit: number): Promise<Product[]> {
    return this.repository
      .createQueryBuilder('product')
      .where('product.status = :status', { status: 'ACTIVE' })
      .orderBy('RANDOM()')
      .limit(limit)
      .getMany();
  }

  private getCategoryAliases(category: string): string[] {
    const normalized = category.trim();
    const upper = normalized.toUpperCase();
    const aliasMap: Record<string, string[]> = {
      'SÁCH': ['SÁCH', 'SACH', 'BOOK', 'Book', 'book'],
      SACH: ['SÁCH', 'SACH', 'BOOK', 'Book', 'book'],
      BOOK: ['SÁCH', 'SACH', 'BOOK', 'Book', 'book'],
      CD: ['CD', 'cd'],
      DVD: ['DVD', 'dvd'],
      'BÁO': ['BÁO', 'BÁO CHÍ', 'BAO', 'BAO CHI', 'NEWSPAPER', 'Newspaper', 'newspaper'],
      'BÁO CHÍ': ['BÁO', 'BÁO CHÍ', 'BAO', 'BAO CHI', 'NEWSPAPER', 'Newspaper', 'newspaper'],
      BAO: ['BÁO', 'BÁO CHÍ', 'BAO', 'BAO CHI', 'NEWSPAPER', 'Newspaper', 'newspaper'],
      'BAO CHI': ['BÁO', 'BÁO CHÍ', 'BAO', 'BAO CHI', 'NEWSPAPER', 'Newspaper', 'newspaper'],
      NEWSPAPER: ['BÁO', 'BÁO CHÍ', 'BAO', 'BAO CHI', 'NEWSPAPER', 'Newspaper', 'newspaper'],
    };

    return Array.from(new Set([normalized, upper, ...(aliasMap[upper] ?? [])]));
  }

  private getMediaTypeForCategory(category: string): string | undefined {
    const upper = category.trim().toUpperCase();
    const mediaTypeMap: Record<string, string> = {
      'SÁCH': 'BOOK',
      SACH: 'BOOK',
      BOOK: 'BOOK',
      CD: 'CD',
      DVD: 'DVD',
      'BÁO': 'NEWSPAPER',
      'BÁO CHÍ': 'NEWSPAPER',
      BAO: 'NEWSPAPER',
      'BAO CHI': 'NEWSPAPER',
      NEWSPAPER: 'NEWSPAPER',
    };

    return mediaTypeMap[upper];
  }

  async countManagerDeleteActions(
    managerId: string,
    start: Date,
    end: Date,
  ): Promise<number> {
    return this.auditLogRepository.count({
      where: {
        performedBy: managerId,
        actionType: In(['DELETE', 'DEACTIVATE']),
        createdAt: Between(start, new Date(end.getTime() - 1)),
      },
    });
  }

  async findAuditLogs(): Promise<ProductLog[]> {
    return this.auditLogRepository.find({
      relations: ['product'],
      order: { createdAt: 'DESC' },
    });
  }

  async saveAuditLog(
    manager: EntityManager,
    params: {
      product: Product;
      actionType: string;
      changedFields: unknown;
      performedBy: string;
      reason?: string;
    },
  ): Promise<void> {
    await manager.getRepository(ProductLog).save({
      product: params.product,
      actionType: params.actionType,
      changedFields: params.changedFields,
      performedBy: params.performedBy,
      reason: params.reason,
    });
  }
}

File: src/backend/src/main/java/vn/aims/product/ProductRepository.java
package vn.aims.product;

import jakarta.persistence.EntityManager;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Optional;
import org.springframework.stereotype.Repository;

/** PostgreSQL-specific search; all user values are bound parameters. */
@Repository
public class ProductRepository {
    private final EntityManager entities;
    public ProductRepository(EntityManager entities) { this.entities = entities; }

    public Optional<Product> findVisible(int id) {
        return entities.createQuery("select p from Product p where p.productID=:id and p.status<>'DELETED'", Product.class)
                .setParameter("id", id).getResultStream().findFirst();
    }

    public <T> T detail(Class<T> type, int id) { return entities.find(type, id); }

    @SuppressWarnings("unchecked")
    public List<Product> random() {
        return entities.createNativeQuery("SELECT * FROM products WHERE status='ACTIVE' ORDER BY RANDOM() LIMIT 20", Product.class)
                .getResultList();
    }

    @SuppressWarnings("unchecked")
    public List<Product> search(ProductSearch filter) {
        StringBuilder sql = new StringBuilder("""
                SELECT DISTINCT p.* FROM products p
                LEFT JOIN books b ON b.product_id=p.product_id
                LEFT JOIN cds c ON c.product_id=p.product_id
                LEFT JOIN cd_tracks t ON t.product_id=p.product_id
                LEFT JOIN dvds d ON d.product_id=p.product_id
                LEFT JOIN newspapers n ON n.product_id=p.product_id
                LEFT JOIN media m ON m.product_id=p.product_id
                WHERE
                """);
        Map<String,Object> params = new LinkedHashMap<>();
        String status = filter.status() == null || filter.status().isEmpty() ? "ACTIVE" : filter.status().toUpperCase(Locale.ROOT);
        sql.append("ALL".equals(status) ? " p.status <> :status" : " p.status = :status");
        params.put("status", "ALL".equals(status) ? "DELETED" : status);
        if (filter.keyword() != null && !ProductSearch.trim(filter.keyword()).isEmpty()) {
            sql.append("""
                     AND (p.title ILIKE :keyword OR p.category ILIKE :keyword OR p.description ILIKE :keyword
                     OR p.barcode ILIKE :keyword OR b.authors ILIKE :keyword OR c.artists ILIKE :keyword
                     OR t.title ILIKE :keyword OR d.director ILIKE :keyword OR n.editor_in_chief ILIKE :keyword
                     OR n.sections ILIKE :keyword OR m.publisher ILIKE :keyword OR m.genre ILIKE :keyword
                     OR m.language ILIKE :keyword)
                    """);
            params.put("keyword", "%" + ProductSearch.trim(filter.keyword()) + "%");
        }
        if (filter.category() != null && !filter.category().isEmpty()) {
            String category = ProductSearch.trim(filter.category());
            String upper = category.toUpperCase(Locale.ROOT);
            String type = switch (upper) {
                case "SÁCH", "SACH", "BOOK" -> "BOOK";
                case "BÁO", "BÁO CHÍ", "BAO", "BAO CHI", "NEWSPAPER" -> "NEWSPAPER";
                case "CD", "DVD" -> upper;
                default -> null;
            };
            List<String> aliases = new ArrayList<>(List.of(category, upper));
            if (type != null) aliases.addAll(switch (type) {
                case "BOOK" -> List.of("SÁCH", "SACH", "BOOK", "Book", "book");
                case "NEWSPAPER" -> List.of("BÁO", "BÁO CHÍ", "BAO", "BAO CHI", "NEWSPAPER", "Newspaper", "newspaper");
                case "CD" -> List.of("CD", "cd");
                default -> List.of("DVD", "dvd");
            });
            sql.append(" AND (p.category IN (:categories)");
            params.put("categories", aliases);
            if (type != null) { sql.append(" OR p.product_type=:categoryType"); params.put("categoryType", type); }
            sql.append(")");
        }
        if (!filter.mediaTypes().isEmpty()) {
            sql.append(" AND p.product_type IN (:types)"); params.put("types", filter.mediaTypes());
        }
        if (filter.minPrice() != null) { sql.append(" AND p.current_price>=:minPrice"); params.put("minPrice", filter.minPrice()); }
        if (filter.maxPrice() != null) { sql.append(" AND p.current_price<=:maxPrice"); params.put("maxPrice", filter.maxPrice()); }
        sql.append(" ORDER BY p.title ASC");
        var query = entities.createNativeQuery(sql.toString(), Product.class);
        params.forEach(query::setParameter);
        return query.getResultList();
    }
}
